src/v/archival/ntp_archiver_service.cc

```diff
@@ -0,0 +1,343 @@
+
+const model::ntp& ntp_archiver::get_ntp() const { return _ntp; }
+
+const ss::lowres_clock::time_point ntp_archiver::get_last_upload_time() const {
+    return _last_upload_time;
+}
+
+const ss::lowres_clock::time_point ntp_archiver::get_last_delete_time() const {
+    return _last_delete_time;
+}
+
+ss::future<bool> ntp_archiver::download_manifest() {
+    gate_guard guard{_gate};
+    auto key = _remote.get_manifest_path();
+    vlog(archival_log.trace, "Download manifest {}", key());
+    auto path = s3::object_key(key());
+    s3::client client(_client_conf);
+    try {
+        auto resp = co_await client.get_object(_bucket, path);
+        co_await _remote.update(resp->as_input_stream());
+    } catch (const s3::rest_error_response& err) {
+        if (err.code() == "NoSuchKey") {
+            // This can happen when we're dealing with new partition for which
+            // manifest wasn't uploaded. But also, this can appen if we uploaded
+            // the first segment and crashed before we were able to upload the
+            // manifest. This shouldn't be the problem though. We will just
+            // re-upload this segment for once.
+            co_return false;
+        }
+        throw;
+    }
+    co_await client.shutdown();
+    co_return true;
+}
+
+ss::future<> ntp_archiver::upload_manifest() {
+    gate_guard guard{_gate};
+    auto key = _remote.get_manifest_path();
+    vlog(archival_log.trace, "Upload manifest {}", key());
+    auto path = s3::object_key(key());
+    auto [is, size] = _remote.serialize();
+    s3::client client(_client_conf);
+    co_await client.put_object(_bucket, path, size, std::move(is));
+    co_await client.shutdown();
+    co_return;
+}
+
+void ntp_archiver::update_local_manifest(storage::log_manager& lm) {
+    auto m = _policy->make_local_manifest(lm);
+    if (m) {
+        _local = *m;
+    } else {
+        throw std::runtime_error("policy error, can't build local manifest");
+    }
+}
+
+const manifest& ntp_archiver::get_local_manifest() const { return _local; }
+
+const manifest& ntp_archiver::get_remote_manifest() const { return _remote; }
+
+/// Returns true if segment matches the metadata
+static bool validate_metadata(
+  const ss::lw_shared_ptr<storage::segment>& segment,
+  const manifest::segment_meta& meta) {
+    // If the metadata don't match the segment then the
+    // local manifest needs to be updated. This can happen when
+    // segment was compacted after updated_local_manifest was
+    // called.
+    //
+    // We shouldn't upload this segment right away because it
+    // might not match the policy.
+    return meta.is_compacted == segment->is_compacted_segment()
+           && meta.size_bytes == segment->size_bytes()
+           && meta.base_offset == segment->offsets().base_offset
+           && meta.committed_offset == segment->offsets().committed_offset;
+}
+
+ss::future<bool> ntp_archiver::upload_segment(
```

Comment:
seems ok to have a bool, but a thought: do we want to return a status::? of sorts. so that something else can compose on this functio. 

i.e: if status::failed_to_upload then you can issue a delete on the s3 api before retrying or smth like that.